multi_camera_parallel_grabber/ex-multi_camera_parallel_grabber.cpp

```diff
@@ -0,0 +1,266 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// ex-multi_camera_parallel_grabber.cpp
+//
+// setup : refer README.md file
+//
+// Prerequisites:
+// *) one ifm 3D camera should be configured to use "Process Interface" for trigger.
+// *) All other camera must be in hardware trigger mode
+// *)
+//
+
+#include <iostream>
+#include <memory>
+#include <chrono>
+#include <vector>
+#include <thread>
+#include <atomic>
+#include <mutex>
+#include <signal.h>
+#include <array>
+#include <tuple>
+#include <ifm3d/camera.h>
+#include <ifm3d/fg.h>
+#include <ifm3d/image.h>
+#include <opencv2/core/core.hpp>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/highgui/highgui.hpp>
+
+namespace
+{   
+  // enum for all the trigger types supported by o3d3xx device
+  enum class trigger_mode : int { FREE_RUN = 1, SW = 2, POSITIVE_EDGE = 3, NEGATIVE_EDGE = 4, POSITIVE_N_NEGATIVE = 5 };
+
+  //flag for stoping the application
+  std::atomic<bool> start = true;
+
+  //Add the IP of cameras to be used for the complete stystem. Address at 0th location must be of camera with software trigger 
+  /* to add device add device address at second last position in the camera_ips vector, increase NUMBER_OF_DEVICES too */
+  const auto NUMBER_OF_DEVICES = 3;
+  const std::array<std::string, NUMBER_OF_DEVICES> camera_ips = { "192.168.0.70", "192.168.0.71","192.168.0.72"};
+
+  auto j_conf = json::parse(R"(
+        {
+          "ifm3d":
+          {
+            "Device":
+            {
+              "ActiveApplication": "1"
+            },
+            "Apps":
+            [
+              {
+                "Index": "1",
+                "TriggerMode":"2",
+                "LogicGraph": "{\"IOMap\": {\"OUT1\": \"RFT\",\"OUT2\": \"AQUFIN\"},\"blocks\": {\"B00001\": {\"pos\": {\"x\": 200,\"y\": 200},\"properties\": {},\"type\": \"PIN_EVENT_IMAGE_ACQUISITION_FINISHED\"},\"B00002\": {\"pos\": {\"x\": 200,\"y\": 75},\"properties\": {},\"type\": \"PIN_EVENT_READY_FOR_TRIGGER\"},\"B00003\": {\"pos\": {\"x\": 600,\"y\": 75},\"properties\": {\"pulse_duration\": 0},\"type\": \"DIGITAL_OUT1\"},\"B00005\": {\"pos\": {\"x\": 600,\"y\": 200},\"properties\": {\"pulse_duration\": 0},\"type\": \"DIGITAL_OUT2\"}},\"connectors\": {\"C00000\": {\"dst\": \"B00003\",\"dstEP\": 0,\"src\": \"B00002\",\"srcEP\": 0},\"C00001\": {\"dst\": \"B00005\",\"dstEP\": 0,\"src\": \"B00001\",\"srcEP\": 0}}}",
+                "Imager":
+                {
+                    "ExposureTime": "1000",
+                    "Type":"under5m_moderate",
+                    "FrameRate":"20"
+                }
+              }
+            ]
+          }
+        }
+    )");
+
+  //for configuration of required trigger on camera
+  void configuration(ifm3d::Camera::Ptr& camera, trigger_mode type)
```

Comment:
use naked pointers for functions which don't participate in object ownership or otherwise don't modify the shared_ptr state (`void configuration(ifm3d::Camera* camera ...`

Reply:
If you disagree, take it up with Herb ;)
https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/